examples/step-55/step-55.cc

```diff
@@ -0,0 +1,1181 @@
+
+    if (solver_type == SolverType::FGMRES_ILU)
+      {
+        TimerOutput::Scope ilu_specific(computing_timer, "(ILU specific)");
+        DoFRenumbering::Cuthill_McKee (dof_handler);
+      }
+
+    DoFRenumbering::component_wise (dof_handler, block_component);
+
+    if (solver_type == SolverType::FGMRES_GMG)
+      {
+        TimerOutput::Scope multigrid_specific(computing_timer, "(Multigrid specific)");
+        TimerOutput::Scope setup_multigrid(computing_timer, "Setup - Multigrid");
+
+        // Distribute only the dofs for velocity finite element
+        velocity_dof_handler.distribute_dofs(velocity_fe);
+
+        // Multigrid only needs the dofs for velocity
+        velocity_dof_handler.distribute_mg_dofs(velocity_fe);
+
+        typename FunctionMap<dim>::type boundary_condition_function_map;
+        BoundaryValuesForVelocity<dim> velocity_boundary_condition;
+        boundary_condition_function_map[0] = &velocity_boundary_condition;
+
+        mg_constrained_dofs.clear();
+        mg_constrained_dofs.initialize(velocity_dof_handler, boundary_condition_function_map);
+        const unsigned int n_levels = triangulation.n_levels();
+
+        mg_interface_matrices.resize(0, n_levels-1);
+        mg_interface_matrices.clear ();
+        mg_matrices.resize(0, n_levels-1);
+        mg_matrices.clear ();
+        mg_sparsity_patterns.resize(0, n_levels-1);
+
+        for (unsigned int level=0; level<n_levels; ++level)
+          {
+            DynamicSparsityPattern csp (velocity_dof_handler.n_dofs(level),
+                                        velocity_dof_handler.n_dofs(level));
+            MGTools::make_sparsity_pattern(velocity_dof_handler, csp, level);
+            mg_sparsity_patterns[level].copy_from (csp);
+
+            mg_matrices[level].reinit(mg_sparsity_patterns[level]);
+            mg_interface_matrices[level].reinit(mg_sparsity_patterns[level]);
+          }
+      }
+
+    {
+      constraints.clear ();
+      // The following makes use of a component mask for interpolation of the boundary values which
+      // is further explained in the vector valued dealii step-20 tutorial
+      DoFTools::make_hanging_node_constraints (dof_handler, constraints);
+      VectorTools::interpolate_boundary_values (dof_handler,
+                                                0,
+                                                Solution<dim>(),
+                                                constraints,
+                                                fe.component_mask(velocities));
+    }
+
+    std::vector<types::global_dof_index> dofs_per_block (2);
+    DoFTools::count_dofs_per_block (dof_handler, dofs_per_block, block_component);
+    const unsigned int n_u = dofs_per_block[0],
+                       n_p = dofs_per_block[1];
+
+    if (solver_type == SolverType::UMFPACK)
+      {
+        TimerOutput::Scope umfpack_specific(computing_timer, "(UMFPACK specific)");
+        constraints.add_line(n_u);
```

Comment:
OK. Add a comment of the form "As discussed in the introduction, we need to fix one degree of freedom of the pressure variable to ensure solvability of the problem. We do this here by marking `n_u` as a constrained degree of freedom. But not making it dependent on any other degree of freedom, this implicitly means that the value of this degree of freedom will be forced to zero."